### PR TITLE
JSON option and metadata in show and scan

### LINKED ISSUE #2459: cylc show: display in parse-able format
A user has requested that we add a functionality for a client command to retrieve parse-able clock trigger information for a task in a running suite. We can probably add an option to `cylc show` to do this.

### PR DESCRIPTION
**Pull request by sadielbartholomew**
(Hopefully) closes  #2459. This is the first piece of working (non-documentation) code I have written on my own so please be kind!

A few points:

1. The specification for issue #2459 was 'functionality... to retrieve parse-able _clock trigger_ information' however Matt and I agreed that it would be best to create a generic ``cylc show`` option which instead returns _all_ task requisite information, to be more comprehensive. The client desiring clock trigger information only could easily manipulate the command output in the shell to extract the relevant information.

2. I believe (could be wrong) this new option will only work if the user provides one or more task IDs and not work for specification of one or more task names given the placement within the 'if task_ids' statement, as for the '--list-prereqs' option. Would it be best to ensure both options work for task names as well as task IDs?

### CODE REVIEW

**Comment by sadielbartholomew** [comment]
I have made some minor edits to my initial PR after verbal feedback from Oliver.

File: bin/cylc-show [hunk 1]
```diff
@@ -46,6 +48,10 @@ def main():
     parser.add_option('--list-prereqs', action="store_true", default=False,
                       help="Print a task's pre-requisites as a list.")
 
+    parser.add_option('--json', action="store_true", default=False,
+                      help="Print all tasks and corresponding requisites in "
+                           "JSON format.")
+
```
Comment (hjoliver) [inline bin/cylc-show]: I think it would be sufficient to say "Print output in JSON format" here.

Comment (sadielbartholomew) [reply bin/cylc-show]: Good suggestions @hjoliver, I will get them implemented on Monday.

**Comment by sadielbartholomew** [comment]
@hjoliver, in my most recent commit added the option to both ``cylc show SUITE`` and ``cylc show SUITE TASK`` i.e. all of the remaining possibilities for user input. Since the testing stage then required three equivalent JSON-format tests I created a new test method in 'test_header' to replace the custom Python test script in my original commit.

**Comment by hjoliver** [comment]
@sadielbartholomew thanks for making those changes; it looks like two tests are now failing as a result though 😬

**Comment by sadielbartholomew** [comment]
Thanks for pointing that out @hjoliver - me and Oliver went through and investigated what would be affected by the change in the scheduler.py functions 'info_get_suite_info' and in 'info_get_task_info' via the 'describe' function in taskdef.py, assuming from this that only really the GUI would be otherwise influenced, but I can see there are a lot of sub-tests failing. Forgive my naivety for making assumptions (always dangerous in software) therefore not running the full test-battery before committing and intending to rely on Travis to pick up on such issues. I will look through the failed tests and see what is going on, adapting them if they are simply reflecting the inclusion of the meta items as I have done for the relevant ``cylc show`` tests and rectifying the code if there are remaining problems. Hopefully that will be fairly straightforward.

**Comment by sadielbartholomew** [comment]
@hjoliver the testing failures should now be resolved and I have renamed the 'descriptions' term as requested, but please hold your review for now as I am going to extend this PR with amendments to ``cylc scan`` so that it also prints out all meta items as it currently does not do so.

**Comment by hjoliver** [comment]
> Forgive my naivety for making assumptions (always dangerous in software) therefore not running the full test-battery before committing and intending to rely on Travis to pick up on such issues. 

Don't worry about that, we all do it (at least I do) - it really doesn't matter so long as all tests pass locally too, before making a release.

**Comment by hjoliver** [comment]
Good idea on `cylc scan --describe`; I'll hold off reviewing for now...

**Comment by sadielbartholomew** [comment]
In my latest commit I have adapted ``cylc scan --description`` to print-out all available metadata (including empty mandatory tag items via '(no <item>)') instead of just the suite title and description or lack thereof. Since I noticed that this option was not explicitly tested
for desired output within the cylc scan tests Oliver and I created some new sub-tests to cover this - this was non-trivial so Oliver kindly talked me through writing code to deal with the user-specific host and port provision issue, unfortunately before I noticed while adapting the 'authentication' tests much later that those tests already contain code to deal with this that we could have just lifted. But credit to @oliver-sanders regardless. A few subtests failed apparently randomly when I ran the battery locally but I have found no reason for it to be related to my code changes so I tentatively await the Travis result. @hjoliver - would you like me to add a JSON option to ``cylc scan`` as for ``cylc show <args>`` given that it already has a parsable outputting option with ``--raw``? I think JSON output could be a useful, complementary, option going forward.

**Comment by hjoliver** [comment]
@sadielbartholomew - thanks for doing this.  Yeah, writing effective tests under our framework is not always easy, to say the least!  Only one test has failed on Travis CI and it looks like an easy fix.   

> would you like me to add a JSON option to ...

Yes, good idea.  It could replace the old --raw option.

**Comment by sadielbartholomew** [comment]
Tests fixed! Just finalising the JSON option and associated testing for ``cylc scan`` and then you can both start reviewing (Hilary when he returns from his holidays I imagine).

**Comment by sadielbartholomew** [comment]
@hjoliver and @oliver-sanders, pre-review changes completed pending a pass from Travis after a local pass. Just to summarise everything I have done to help you navigate the amended files:

1. Added a '--json' option to ``cylc show`` and ``cylc scan`` providing the standard output information in JSON for each case;
2. Created a new test 'cmp_json_ok' to the 'test_header' to do an output <-> expected output comparison whilst checking for valid JSON, and applied it to test the--json outputs;
3. Created further new tests for the ``cylc scan --describe`` and ``--raw`` outputs (in 'tests/authentication' along with the other ``scan`` tests) equivalent to those already in place for ``--full``, as explicit tests for these were missing, and made a minor change to ``--raw`` to remove potential ugly whitespace in the middle of multi-line descriptions;
4. Revised ``cylc show`` and ``cylc scan --describe`` to output all metadata instead of just the title and description (version-permitting) and changed relevant help- and doc-strings accordingly.

**Comment by sadielbartholomew** [comment]
Thanks @oliver-sanders, well spotted in each case and I will get a commit in shortly making the relevant amendments and also as I proposed in person add extra levels into the JSON for state totals, e.g. rather than:
...
	"states": {
		"1": "failed:1 waiting:1"
	},
...
create a more useful parsable structure of:
...
	"states": {
		"1": {
                      "failed": 1,
                       "waiting": 1
                       }
	}
...
with the number 'key' a string but the frequency 'values' integers, as I believe you suggested was best in context.
In terms of pretty-printing the JSON with 'indent=4' I will do this for ``cylc show`` as well as ``cylc scan`` for consistency unless you can suggest any reason not to.

File: bin/cylc-scan [hunk 2]
```diff
@@ -237,10 +249,14 @@ def main():
                     # KEY_UPDATE_TIME and Compat:<=7.5.0
                     value = suite_identity.get(key)
                 if value:
+                    # Remove new-line characters for readability
+                    value_to_output = str(value).replace("\n", " ")
+                    # Remove whitespace from multiline indents for readability
+                    while "  " in value_to_output:
+                        value_to_output = value_to_output.replace("  ", " ")
```
[the remark this answers is not in the document]
Comment (sadielbartholomew) [reply bin/cylc-scan]: This is really nice! And it just fits within the line length limit given the necessary indentation which is quite satisfying.

**Comment by sadielbartholomew** [comment]
I was trying to implement the changes based on Oliver's feedback above today but whilst most were trivial there has been a complication as we agreed upon my suggestion that suite names would be better as the primary key in the Python dictionary to dump instead of simply a JSON-structured 'list' of all suites and this means that the suite name pattern matching to use to filter non-test suites in the authentication i.e. scan tests no longer work as all suites including those to filter out go into one structure. I am nearly there at finding a way round this but I am taking Christmas leave now. Will resolve and commit upon my return in 2018.

**Comment by sadielbartholomew** [comment]
In my most recent commit I have applied @oliver-sanders' suggestions as above and ensured extraction of only pattern-matched suites with the pattern option as required for testing (see my previous comment). @hjoliver and @matthewrmshin that is my final commit until you are able to review. Thanks all.

**Comment by sadielbartholomew** [comment]
Oh dear, will squash in the necessary fixes to get the 'syntax/01-pycodestyle.t' test to pass tomorrow. Otherwise done until review.

File: bin/cylc-scan [hunk 3]
```diff
@@ -253,6 +266,36 @@ def main():
...
+                    # KEY_UPDATE_TIME and Compat:<=7.5.0
+                    value = suite_identity.get(key)
+                if value:
+                    suite_inf.update({key: ' '.join([x.strip() for x in
+                                      str(value).split('\n') if x])})
+                totals = suite_identity.get(KEY_STATES)
+                if totals is None:
+                    continue
+                point_state_lines = get_point_state_count_lines(
+                    *totals, use_color=options.color)
```
Comment (oliver-sanders) [inline bin/cylc-scan]: er?

Comment (sadielbartholomew) [reply bin/cylc-scan]: Nice spot! I did ensure ``--json``  cannot be used with the ``--color`` option on lines 185-188 (in case anyone wanted to check) but I agree it is best to just remove this argument as it is therefore unnecessary and we can then trim the line count by 1. Will update in a minute.

File: bin/cylc-scan [hunk 4]
```diff
@@ -253,6 +266,35 @@ def main():
...
+                    state_dict = {}
+                    if point:
+                        for state_inf in state_line.split(' '):
+                            if ':' in state_inf:
+                                key, value = state_inf.split(':')
+                                try:
+                                    state_dict.update({key: int(value)})
+                                except (TypeError, ValueError):
+                                    state_dict.update({key: str(value)})
+                        suite_inf.update({KEY_STATES: {point: state_dict}})
```
Comment (hjoliver) [inline bin/cylc-scan]: Dict update here will replace the value of `suite_inf[KEY_STATES]` each time through the loop, so this code only retains the state summary for the last `point`.

**Comment by hjoliver** [comment]
So we need `cylc scan` tests with multiple cycle points, to cover this.

File: bin/cylc-scan [hunk 5]
```diff
@@ -253,6 +266,35 @@ def main():
...
+            json_dict[name] = suite_inf
+            continue
```
[the remark this answers is not in the document]
Comment (sadielbartholomew) [reply bin/cylc-scan]: No problem, starting to amend all these points now - to clarify for this point 294 and 295 aren't in that loop but 280:294 are (I think you meant to put 294 rather than 296) so I will reduce the indentation level by 1 for that block as suggested.

File: bin/cylc-scan (hunk 4 above)
Comment (sadielbartholomew) [reply bin/cylc-scan]: Lesson learnt regarding the need for comprehensive tests!

**Comment by sadielbartholomew** [comment]
TO DO
Hilary's corrections are now included, except for multiple cycle points in the tests which I will add shortly. The '01-description.t' test is giving a blank JSON structure when it should not after the reduction in indentation level of the block 280:296 so I need to rectify that, and I can definitely increase the efficiency of the changes in my previous commit, but commiting for now while they are at least functional.

File: bin/cylc-scan (hunk 3 above)
Comment (sadielbartholomew) [reply bin/cylc-scan]: I also spotted this in the equivalent line within the ``--raw`` option encompassing statement and removed that too in my recent commit, as colour shouldn't (though strictly can't anyway, via parser error) be used with that either.

**Comment by sadielbartholomew** [comment]
@hjoliver & @oliver-sanders I have now modified ``cylc scan`` and its tests in the authentication test directory, including the 'basic' suite and 'override' suite graphing in order to include a second cycle point, to resolve, as far as reckon at least, all the issues you raised. Please re-review at your convenience. @oliver-sanders changing part of line 36 in 'tests/authentication/04-shutdown.t' from ``--task=foo --status=failed --point=1`` to ``--task=pub --status=succeeded --point=2`` as you suggested would make that test stronger kept resulting in either of the 'validate' or 'stop' (first and last) subtests on that test only failing for a reason I could not work out - hopefully we can discuss possible reasons for this next week.

**Comment by sadielbartholomew** [comment]
Actually, is it worth combining the try/except statements to extract the metadata from the ``--raw`` (lines 246-254) and ``--json`` (lines 273-281) options into an 'if options.raw_format or options.json_format' (even 'or options.describe' too though this deals with the  KEY_TITLE and KEY_DESCRIPTION separately so might ) statement on line 243 before the individual option statements to avoid a significant amount of repetition? Or is it best to leave the code for each option segregated for clarity?

**Comment by sadielbartholomew** [comment]
Oliver says 'yes' (if I feel up to it) RE my previous comment question so I am going to do a bit (maybe a lot) of restructuring of the code in a new commit.

**Comment by sadielbartholomew** [comment]
Refactoring complete pending Travis CI, at least to a suitable level for my liking - maybe not for the reviewers' liking! One point: do we want to include 'KEY_UPDATE_TIME' in the output print-out from ``--raw`` and ``--json``? I assumed not and left it out but could easily add it in, though I would need to create some sort of test able to check the update time was any float as I imagine update time will vary in every case.

**Comment by matthewrmshin** [comment]
(Testing this now.)
